File: unionfind.py
class UnionFind:
    def __init__(self, n: int):
        self.par = [i for i in range(n)]
        self.rank = [1] * n
    
    def find(self, i: int) -> int:
        if self.par[i] == i:
            return i
        return self.find(self.par[i])

    def union(self, i: int, j: int) -> bool:
        p1 = self.find(i)
        p2 = self.find(j)
        if p1 == p2:
            return False
        
        if self.rank[p1] >= self.rank[p2]:
            self.par[p2] = p1
            self.rank[p1] += self.rank[p2]
        else:
            self.par[p1] = p2
            self.rank[p2] += self.rank[p1]
        return True

#use this
class UnionFind:
    def __init__(self, n: int):
        self._fa = list(range(n))
        self._size = [1] * n
        self.cc = n

    def find(self, x: int) -> int:
        if self._fa[x] != x:
            self._fa[x] = self.find(self._fa[x])
        return self._fa[x]

    def is_same(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def merge(self, from_: int, to: int) -> bool:
        x, y = self.find(from_), self.find(to)
        if x == y:
            return False
        self._fa[x] = y
        self._size[y] += self._size[x]
        self.cc -= 1
        return True

    def get_size(self, x: int) -> int:
        return self._size[self.find(x)]
    
    def init(self):
        self._fa = list(range(self.cc))
        self._size = [1] * self.cc

File: go_templates/UnionFind.go
package main

type UnionFind struct {
	parent []int
	rank   []int
	size   int
}

func UF(n int) *UnionFind {
	uf := &UnionFind{
		parent: make([]int, n),
		rank:   make([]int, n),
		size:   n,
	}
	for i := range n {
		uf.parent[i] = i
		uf.rank[i] = 1
	}
	return uf
}

func (uf *UnionFind) Find(i int) int {
	for uf.parent[i] != i {
		uf.parent[i] = uf.parent[uf.parent[i]]
		i = uf.parent[i]
	}
	return i
}

func (uf *UnionFind) Merge(a, b int) bool {
	pa := uf.Find(a)
	pb := uf.Find(b)
	if pa == pb {
		return false
	}
	if uf.rank[pa] > uf.rank[pb] {
		uf.parent[pb] = pa
	} else if uf.rank[pb] > uf.rank[pa] {
		uf.parent[pa] = pb
	} else {
		uf.parent[pb] = pa
		uf.rank[pa]++
	}
	uf.size--
	return true
}

func (uf *UnionFind) Size() int {
	return uf.size
}
